File: solutions/csharp/pig-latin/1/PigLatin.cs
public static class PigLatin
{
    public static string Translate(string word) =>
        string.Join(" ", word.Split(' ').Select(TranslateWord));

    private static string TranslateWord(string word) =>
        StartsWithVowelSound(word) ? $"{word}ay" : TranslateConsonant(word);

    private static string TranslateConsonant(string word)
    {
        bool IsClusterEnd(int index) =>
            index > 0 && word[index] == 'y' || IsVowel(word[index]) || IsQu(word, index);

        var index = Enumerable.Range(0, word.Length)
            .First(IsClusterEnd);

        var split = IsQu(word, index) ? index + 2 : index;

        return $"{word[split..]}{word[..split]}ay";
    }

    private static bool StartsWithVowelSound(string word) =>
        IsVowel(word[0]) || word.StartsWith("xr") || word.StartsWith("yt");

    private static bool IsVowel(char character) => "aeiou".Contains(character);

    private static bool IsQu(string word, int index) =>
        word[index] == 'q' && index + 1 < word.Length && word[index + 1] == 'u';
}

File: solutions/kotlin/pig-latin/1/src/main/kotlin/PigLatin.kt
object PigLatin {
    fun translate(phrase: String): String =
        phrase
            .split(" ")
            .joinToString(" ", transform = ::translateWord)

    private fun translateWord(word: String): String =
        if (startsWithVowelSound(word)) {
            "${word}ay"
        } else {
            translateConsonantWord(word)
        }

    private fun translateConsonantWord(word: String): String {
        val split = word.indices.first { index -> isClusterEnd(word, index) }
        val clusterEnd = if (isQu(word, split)) split + 2 else split

        return "${word.drop(clusterEnd)}${word.take(clusterEnd)}ay"
    }

    private fun startsWithVowelSound(word: String): Boolean =
        word.first() in VOWELS || word.startsWith("xr") || word.startsWith("yt")

    private fun isClusterEnd(word: String, index: Int,): Boolean =
        index > 0 && word[index] == 'y' || word[index] in VOWELS || isQu(word, index)

    private fun isQu(word: String, index: Int,): Boolean =
        word[index] == 'q' && word.getOrNull(index + 1) == 'u'

    private val VOWELS = setOf('a', 'e', 'i', 'o', 'u')
}
